src/storage/src/hummock/file_cache/store.rs

```diff
@@ -12,431 +12,533 @@
-                .store
-                .meta_file
-                .write()
-                .instrument(tracing::trace_span!("meta_write_lock_free_slots"))
-                .await;
-            for slot in freelist {
-                if let Some(bloc) = guard.free(slot) {
-                    let offset = bloc.bidx as u64 * self.block_size as u64;
-                    let len = bloc.blen(self.block_size as u32) as usize;
-                    self.store.cache_file.punch_hole(offset, len)?;
-                }
-            }
-            drop(guard);
-        }
-
-        if self.is_empty() {
-            return Ok((vec![], vec![]));
-        }
-
-        // Write new cache entries.
-
-        let mut slots = Vec::with_capacity(self.blocs.len());
-
-        let mut boff: Option<u32> = None;
-
-        for buffer in self.buffers {
-            let blen = buffer.len();
-
-            let timer = self.store.metrics.disk_write_latency.start_timer();
-            let offset = self.store.cache_file.append(buffer).await? / self.block_size as u64;
-            timer.observe_duration();
-
-            if boff.is_none() {
-                boff = Some(offset.try_into().unwrap());
-            }
-
-            self.store.metrics.disk_write_bytes.inc_by(blen as f64);
-            self.store.metrics.disk_write_io_size.observe(blen as f64);
-        }
-
-        let boff = boff.unwrap();
+impl Value for Box<Block> {
+    fn serialized_len(&self) -> usize {
+        self.raw_data().len()
+    }
 
-        for bloc in &mut self.blocs {
-            bloc.bidx += boff;
-        }
+    fn write(&self, mut buf: &mut [u8]) {
+        buf.put_slice(self.raw_data())
+    }
 
-        // Write guard is only needed when updating meta file, for data file is append-only.
-        let mut guard = self
-            .store
-            .meta_file
-            .write()
-            .instrument(tracing::trace_span!("meta_write_lock_update_slots"))
-            .await;
+    fn read(buf: &[u8]) -> Self {
+        let data = Bytes::copy_from_slice(buf);
+        let block = Block::decode_from_raw(data);
+        Box::new(block)
+    }
+}
 
-        for (key, bloc) in self.keys.iter().zip_eq_fast(self.blocs.iter()) {
-            slots.push(guard.insert(key, bloc)?);
-        }
+impl Value for Box<Sstable> {
+    fn weight(&self) -> usize {
+        std::mem::size_of::<Self>()
+    }
 
-        Ok((self.keys, slots))
+    fn serialized_len(&self) -> usize {
+        8 + self.meta.encoded_size() // id (8B) + meta size
     }
-}
 
-pub struct StoreOptions {
-    pub dir: String,
-    pub capacity: usize,
-    pub buffer_capacity: usize,
-    pub cache_file_fallocate_unit: usize,
-    pub cache_meta_fallocate_unit: usize,
-    pub cache_file_max_write_size: usize,
+    fn write(&self, mut buf: &mut [u8]) {
+        buf.put_u64(self.id);
+        // TODO(MrCroxx): avoid buffer copy
+        let mut buffer = vec![];
+        self.meta.encode_to(&mut buffer);
```

Comment:
Is it possible to encode to buf directly?

Reply:
I think it is. But it requires a lot of modification on `Block`. I'll do it in the following PRs.

Reply:
I think we can try modifying the `encode_to` method of `SstableMeta` from `buf: &mut Vec<u8>` to `buf: impl BufMut`.

Reply:
+1 .  Thought the same as you.